### ISSUE TITLE
Node constraints consistently cannot be found

### ISSUE THREAD

**Issue by mrapczynski** [docker/swarm#1563]
I set up Swarm for the first time using the latest 1.0.1, and getting a container going with a node constraint is not working. I've poured over the docs, examples, other issues here in the repo, looked for spelling mistakes - I don't get what I am missing - driving me crazy.

Constantly getting this error whether I try docker run or try to build a container using compose.

`Error response from daemon: unable to find a node that satisfies environment==management`

Here is what `docker info` reports. Please note that each node has the appropriate daemon label applied.

```
[docker@dockermgmt ~]$ docker info
Containers: 6
Images: 6
Role: replica
Primary: dockern1.ad.fhda.edu:3376
Strategy: spread
Filters: health, port, dependency, affinity, constraint
Nodes: 3
 dockermgmt.ad.fhda.edu: dockermgmt.ad.fhda.edu:2376
  └ Status: Healthy
  └ Containers: 2
  └ Reserved CPUs: 0 / 4
  └ Reserved Memory: 0 B / 3.888 GiB
  └ Labels: environment=management, executiondriver=native-0.2, kernelversion=3.10.0-327.3.1.el7.x86_64, operatingsystem=CentOS Linux 7 (Core), storagedriver=devicemapper
 dockern1.ad.fhda.edu: dockern1.ad.fhda.edu:2376
  └ Status: Healthy
  └ Containers: 2
  └ Reserved CPUs: 0 / 8
  └ Reserved Memory: 0 B / 8.022 GiB
  └ Labels: environment=production, executiondriver=native-0.2, kernelversion=3.10.0-327.3.1.el7.x86_64, operatingsystem=CentOS Linux 7 (Core), storagedriver=devicemapper
 dockertest2.ad.fhda.edu: dockertest2.ad.fhda.edu:2376
  └ Status: Healthy
  └ Containers: 2
  └ Reserved CPUs: 0 / 8
  └ Reserved Memory: 0 B / 8.022 GiB
  └ Labels: environment=test, executiondriver=native-0.2, kernelversion=3.10.0-327.3.1.el7.x86_64, operatingsystem=CentOS Linux 7 (Core), storagedriver=devicemapper
CPUs: 20
Total Memory: 19.93 GiB
Name: 8e6008ca47f1
```

So I try to create a container as follows:

```
[docker@dockermgmt ~]$ docker run -d --name registry -e "constraint:environment==management" -p 5000:5000 registry-fhda:2.2.1
Error response from daemon: unable to find a node that satisfies environment==management
```

...and the error comes up. What am I doing wrong?

**Comment by abronan** [comment]
Hi @mrapczynski,

Indeed that's weird, just double-checked with a fresh cluster and I can use constraints just fine.. I see that you are using **Manager HA** talking to a `Replica`, can you try to contact the `Primary` and see if this outputs the same informations on `docker info`? It should but just to double check.

Most likely it seems to be a synchronization issue on labels. At a first glance what seems to happen is that you talk to the `Replica` which sees the labels, but the `Primary` still has an *old* view of the labels. The `Replica` just passes on the request to the `Primary` which then fails because it does not know about the `environment` label. By any chance did you restart `docker` during the process off setting up your cluster?

Also does this work when you put a constraint on the `executiondriver` or the `kernelversion`?

**Comment by mrapczynski** [comment]
Hi @abronan. Thank you for the follow up.

Up front, it's now working, but I'm not certain how I got to that point so I'll go through the details below.

Yes, I could confirm at that time that the primary and both replicas all had the same label information each time I tried `docker info`.

I had restarted the replica multiple times, and but only `swarm-manager` on the primary each time I tweaked a daemon label, and that seemed to do fine. Please let me know if that is incorrect. Initially I was not sure if I was doing the label wrong or it was a syntax error, and so I tried different permutations of labels like `environment` and then namespacing it such as `fhda.environment`. Therefore many restarts. I also tried both `constraint:environment==management` and `affinity:environment==management`.

Now, let me run something else by you that caught my attention. Yesterday after I  wrote up this issue, I was a stopping point and started cleaning up my windows and saving stuff I had pasted in Atom. Below is how I have my three swarm clients set up:

```
#dockermgmt
docker run -d --name swarm-client swarm join --addr=dockermgmt.ad.fhda.edu:2376 consul://dockermgmt.ad.fhda.edu:8500/swarm

#dockern1
docker run -d --name swarm-client swarm join --addr=dockern1.ad.fhda.edu:2376 consul://dockern1.ad.fhda.edu:8500/swarm

#dockertest2
docker run -d --name swarm-client swarm join --addr=dockertest2.ad.fhda.edu:2376 consul://dockertest2.ad.fhda.edu:8500/swarm
```

I noticed only then that I had set up the client on dockern1 (also the swarm primary) incorrectly. It was not incorrect in the sense that it would not work, it was just not ideal. Each Docker server is a member of a Consul cluster, and in turn these three servers are running manager HA. I will be joining two additional servers as clients only next week.

The example above is my corrected one, but originally I had set up the client as follows: `docker run -d --name swarm-client swarm join --addr=dockern1.ad.fhda.edu:2376 consul://dockermgmt.ad.fhda.edu:8500/swarm` - notice that dockern1 client would interface with the Consul client on the other server. In the correct verison, I want the client to talk to its own Consul client on the same host.

I fixed that by redeploying the client, and then I _fully restarted the Docker daemon_ on the primary. Boom! Suddenly the constraints started work when I gave it once last try.

So... this is where I am at. Was fully restarting the daemon on the primary the actual fix? What puzzles me is before all that `docker info` still showed the correct labels so it's not like the managers did not have the latest info. I wonder if I ran into #1281 and just did not realize it at that moment. I did all of the exact steps mentioned in the intro comment on that issue.

**Comment by abronan** [comment]
@mrapczynski Thank you for the update

I think that indeed you might have ran into #1281 as the labels are not a mechanism linked to Consul or HA in any way. Still confused about why the Primary reported the label in the first place (it shouldn't) but maybe it missed the label somehow at startup. (this should be fixed by #1370 which refreshes the labels and allows to update those labels if an engine restarts).

Will close this one as a duplicate but feel free to comment on the linked issue. Thanks for the time invested into investigating this issue! Much appreciated.
